<!DOCTYPE html>
<html lang="en"><head>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.5.57">

  <meta name="author" content="Professor: Davi Moreira">
  <title>MGMT 47400: Predictive Analytics –  MGMT 47400: Predictive Analytics </title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title"><span style="font-size: 100%;"> MGMT 47400: Predictive Analytics </span></h1>
  <p class="subtitle"><span style="font-size: 150%;"> Unsupervised Learning </span></p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Professor: Davi Moreira 
</div>
</div>
</div>

</section>
<section id="overview" class="slide level2 center">
<h2>Overview</h2>
<div>
<div class="columns">
<div class="column" style="text-align: center; justify-content: center; align-items: center;">
<ul>
<li>XXXX</li>
</ul>
</div><div class="column" style="text-align: center; justify-content: center; align-items: center;">
<ul>
<li>XXX</li>
</ul>
</div></div>
</div>
</section>
<section>
<section id="xxx" class="title-slide slide level1 center" data-background-color="#cfb991">
<h1>XXX</h1>

</section>
<section id="unsupervised-learning" class="slide level2 center">
<h2>Unsupervised Learning</h2>
<p><strong>Unsupervised vs Supervised Learning:</strong></p>
<ul>
<li class="fragment"><p>Most of this course focuses on <em>supervised learning</em> methods such as regression and classification.</p></li>
<li class="fragment"><p>In that setting we observe both a set of features <span class="math inline">\(X_1, X_2, \ldots, X_p\)</span> for each object, as well as a response or outcome variable <span class="math inline">\(Y\)</span>. The goal is then to predict <span class="math inline">\(Y\)</span> using <span class="math inline">\(X_1, X_2, \ldots, X_p\)</span>.</p></li>
<li class="fragment"><p>Here we instead focus on <em>unsupervised learning</em>, where we observe only the features <span class="math inline">\(X_1, X_2, \ldots, X_p\)</span>. We are not interested in prediction, because we do not have an associated response variable <span class="math inline">\(Y\)</span>.</p></li>
</ul>
</section>
<section id="the-goals-of-unsupervised-learning" class="slide level2 center">
<h2>The Goals of Unsupervised Learning</h2>
<ul>
<li class="fragment"><p>The goal is to discover interesting things about the measurements: is there an informative way to visualize the data? Can we discover subgroups among the variables or among the observations?</p></li>
<li class="fragment"><p>We discuss two methods:</p>
<ul>
<li class="fragment"><p><em>principal components analysis</em>, a tool used for data visualization or data pre-processing before supervised techniques are applied, and</p></li>
<li class="fragment"><p><em>clustering</em>, a broad class of methods for discovering unknown subgroups in data.</p></li>
</ul></li>
</ul>
</section>
<section id="the-challenge-of-unsupervised-learning" class="slide level2 center">
<h2>The Challenge of Unsupervised Learning</h2>
<ul>
<li class="fragment"><p>Unsupervised learning is more subjective than supervised learning, as there is no simple goal for the analysis, such as prediction of a response.</p></li>
<li class="fragment"><p>But techniques for unsupervised learning are of growing importance in a number of fields:</p>
<ul>
<li class="fragment">subgroups of breast cancer patients grouped by their gene expression measurements,</li>
<li class="fragment">groups of shoppers characterized by their browsing and purchase histories,</li>
<li class="fragment">movies grouped by the ratings assigned by movie viewers.</li>
</ul></li>
</ul>
</section>
<section id="another-advantage" class="slide level2 center">
<h2>Another advantage</h2>
<ul>
<li class="fragment"><p>It is often easier to obtain <em>unlabeled data</em> — from a lab instrument or a computer — than <em>labeled data</em>, which can require human intervention.</p></li>
<li class="fragment"><p>For example, it is difficult to automatically assess the overall sentiment of a movie review: is it favorable or not?</p></li>
</ul>
</section>
<section id="principal-components-analysis" class="slide level2 center">
<h2>Principal Components Analysis</h2>
<ul>
<li class="fragment">PCA produces a low-dimensional representation of a dataset. It finds a sequence of linear combinations of the variables that have maximal variance, and are mutually uncorrelated.</li>
<li class="fragment">Apart from producing derived variables for use in supervised learning problems, PCA also serves as a tool for data visualization.</li>
</ul>
</section>
<section id="principal-components-analysis-details" class="slide level2 center">
<h2>Principal Components Analysis: details</h2>
<ul>
<li class="fragment">The <em>first principal component</em> of a set of features <span class="math inline">\(X_1, X_2, \ldots, X_p\)</span> is the normalized linear combination of the features</li>
</ul>
<p><span class="math display">\[
  Z_1 = \phi_{11}X_1 + \phi_{21}X_2 + \ldots + \phi_{p1}X_p
\]</span></p>
<p>that has the largest variance. By <em>normalized</em>, we mean that <span class="math inline">\(\sum_{j=1}^p \phi_{j1}^2 = 1\)</span>.</p>
<ul>
<li class="fragment">We refer to the elements <span class="math inline">\(\phi_{11}, \ldots, \phi_{p1}\)</span> as the loadings of the first principal component; together, the loadings make up the principal component loading vector,</li>
</ul>
<p><span class="math display">\[
  \phi_1 = (\phi_{11}, \phi_{21}, \ldots, \phi_{p1})^T.
\]</span></p>
<ul>
<li class="fragment">We constrain the loadings so that their sum of squares is equal to one, since otherwise setting these elements to be arbitrarily large in absolute value could result in an arbitrarily large variance.</li>
</ul>
</section>
<section id="pca-example" class="slide level2 center">
<h2>PCA: example</h2>

<img data-src="figs/12_1_1-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"><p>The population size (<em>pop</em>) and ad spending (<em>ad</em>) for 100 different cities are shown as purple circles. The green solid line indicates the first principal component direction, and the blue dashed line indicates the second principal component direction.</p>
</section>
<section id="computation-of-principal-components" class="slide level2 center">
<h2>Computation of Principal Components</h2>
<ul>
<li class="fragment"><p>Suppose we have a <span class="math inline">\(n \times p\)</span> data set <span class="math inline">\(\mathbf{X}\)</span>. Since we are only interested in variance, we assume that each of the variables in <span class="math inline">\(\mathbf{X}\)</span> has been centered to have mean zero (that is, the column means of <span class="math inline">\(\mathbf{X}\)</span> are zero).</p></li>
<li class="fragment"><p>We then look for the linear combination of the sample feature values of the form <span class="math display">\[
z_{i1} = \phi_{11}x_{i1} + \phi_{21}x_{i2} + \ldots + \phi_{p1}x_{ip} \tag{1}
\]</span></p>
<p>for <span class="math inline">\(i = 1, \ldots, n\)</span> that has the largest sample variance, subject to the constraint that <span class="math inline">\(\sum_{j=1}^p \phi_{j1}^2 = 1\)</span>.</p></li>
<li class="fragment"><p>Since each of the <span class="math inline">\(x_{ij}\)</span> has mean zero, then so does <span class="math inline">\(z_{i1}\)</span> (for any values of <span class="math inline">\(\phi_{j1}\)</span>). Hence the sample variance of the <span class="math inline">\(z_{i1}\)</span> can be written as</p></li>
</ul>
<p><span class="math display">\[
  \frac{1}{n} \sum_{i=1}^n z_{i1}^2.
\]</span></p>
</section>
<section id="computation-continued" class="slide level2 center">
<h2>Computation: continued</h2>
<ul>
<li class="fragment">Plugging in (1) the first principal component loading vector solves the optimization problem</li>
</ul>
<p><span class="math display">\[
  \text{maximize}_{\phi_{11}, \ldots, \phi_{p1}} \frac{1}{n} \sum_{i=1}^n \left( \sum_{j=1}^p \phi_{j1} x_{ij} \right)^2 \quad \text{subject to} \quad \sum_{j=1}^p \phi_{j1}^2 = 1.
\]</span></p>
<ul>
<li class="fragment"><p>This problem can be solved via a singular-value decomposition of the matrix <span class="math inline">\(\mathbf{X}\)</span>, a standard technique in linear algebra.</p></li>
<li class="fragment"><p>We refer to <span class="math inline">\(Z_1\)</span> as the first principal component, with realized values <span class="math inline">\(z_{11}, \ldots, z_{n1}\)</span>.</p></li>
</ul>
</section>
<section id="geometry-of-pca" class="slide level2 center">
<h2>Geometry of PCA</h2>
<ul>
<li class="fragment"><p>The loading vector <span class="math inline">\(\phi_1\)</span> with elements <span class="math inline">\(\phi_{11}, \phi_{21}, \ldots, \phi_{p1}\)</span> defines a direction in feature space along which the data vary the most.</p></li>
<li class="fragment"><p>If we project the <span class="math inline">\(n\)</span> data points <span class="math inline">\(x_1, \ldots, x_n\)</span> onto this direction, the projected values are the principal component scores <span class="math inline">\(z_{11}, \ldots, z_{n1}\)</span> themselves.</p></li>
</ul>
</section>
<section id="further-principal-components" class="slide level2 center">
<h2>Further principal components</h2>
<ul>
<li class="fragment"><p>The second principal component is the linear combination of <span class="math inline">\(X_1, \ldots, X_p\)</span> that has maximal variance among all linear combinations that are <em>uncorrelated</em> with <span class="math inline">\(Z_1\)</span>.</p></li>
<li class="fragment"><p>The second principal component scores <span class="math inline">\(z_{12}, z_{22}, \ldots, z_{n2}\)</span> take the form</p></li>
</ul>
<p><span class="math display">\[
  z_{i2} = \phi_{12}x_{i1} + \phi_{22}x_{i2} + \ldots + \phi_{p2}x_{ip},
\]</span></p>
<p>where <span class="math inline">\(\phi_2\)</span> is the second principal component loading vector, with elements <span class="math inline">\(\phi_{12}, \phi_{22}, \ldots, \phi_{p2}\)</span>.</p>
</section>
<section id="further-principal-components-continued" class="slide level2 center">
<h2>Further principal components: continued</h2>
<ul>
<li class="fragment"><p>It turns out that constraining <span class="math inline">\(Z_2\)</span> to be uncorrelated with <span class="math inline">\(Z_1\)</span> is equivalent to constraining the direction <span class="math inline">\(\phi_2\)</span> to be orthogonal (perpendicular) to the direction <span class="math inline">\(\phi_1\)</span>. And so on.</p></li>
<li class="fragment"><p>The principal component directions <span class="math inline">\(\phi_1, \phi_2, \phi_3, \ldots\)</span> are the ordered sequence of right singular vectors of the matrix <span class="math inline">\(\mathbf{X}\)</span>, and the variances of the components are <span class="math inline">\(\frac{1}{n}\)</span> times the squares of the singular values. There are at most <span class="math inline">\(\min(n - 1, p)\)</span> principal components.</p></li>
</ul>
</section>
<section id="illustration" class="slide level2 center">
<h2>Illustration</h2>
<ul>
<li class="fragment"><p><strong>USArrests</strong> data: For each of the fifty states in the United States, the data set contains the number of arrests per 100,000 residents for each of three crimes: <strong>Assault</strong>, <strong>Murder</strong>, and <strong>Rape</strong>. We also record <strong>UrbanPop</strong> (the percent of the population in each state living in urban areas).</p></li>
<li class="fragment"><p>The principal component score vectors have length <span class="math inline">\(n = 50\)</span>, and the principal component loading vectors have length <span class="math inline">\(p = 4\)</span>.</p></li>
<li class="fragment"><p>PCA was performed after standardizing each variable to have mean zero and standard deviation one.</p></li>
</ul>
</section>
<section id="usarrests-data-pca-plot" class="slide level2 center">
<h2>USArrests data: PCA plot</h2>

<img data-src="figs/12_1-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"><p>The PCA plot shows the first and second principal components for the USArrests dataset, with arrows indicating the loadings for the variables <strong>UrbanPop</strong>, <strong>Rape</strong>, <strong>Assault</strong>, and <strong>Murder</strong>. State names are displayed based on their principal component scores.</p>
</section>
<section id="figure-details" class="slide level2 center">
<h2>Figure details</h2>
<p>The first two principal components for the USArrests data.</p>
<ul>
<li class="fragment"><p>The blue state names represent the scores for the first two principal components.</p></li>
<li class="fragment"><p>The orange arrows indicate the first two principal component loading vectors (with axes on the top and right). For example, the loading for <strong>Rape</strong> on the first component is 0.54, and its loading on the second principal component is 0.17 [the word <strong>Rape</strong> is centered at the point (0.54, 0.17)].</p></li>
<li class="fragment"><p>This figure is known as a <strong>biplot</strong>, because it displays both the principal component scores and the principal component loadings.</p></li>
</ul>
</section>
<section id="pca-loadings" class="slide level2 center">
<h2>PCA loadings</h2>
<table class="caption-top">
<thead>
<tr class="header">
<th></th>
<th>PC1</th>
<th>PC2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Murder</td>
<td>0.5358995</td>
<td>-0.4181809</td>
</tr>
<tr class="even">
<td>Assault</td>
<td>0.5831836</td>
<td>-0.1879856</td>
</tr>
<tr class="odd">
<td>UrbanPop</td>
<td>0.2781909</td>
<td>0.8728062</td>
</tr>
<tr class="even">
<td>Rape</td>
<td>0.5434321</td>
<td>0.1673186</td>
</tr>
</tbody>
</table>
</section>
<section id="another-interpretation-of-principal-components" class="slide level2 center">
<h2>Another Interpretation of Principal Components</h2>

<img data-src="figs/12_1_2-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"><p>The left plot shows a 3D visualization with projections onto the first two principal components, while the right plot displays the data in 2D using the first and second principal components.</p>
</section>
<section id="pca-find-the-hyperplane-closest-to-the-observations" class="slide level2 center">
<h2>PCA find the hyperplane closest to the observations</h2>
<ul>
<li class="fragment"><p>The first principal component loading vector has a very special property: it defines the line in <span class="math inline">\(p\)</span>-dimensional space that is <em>closest</em> to the <span class="math inline">\(n\)</span> observations (using average squared Euclidean distance as a measure of closeness).</p></li>
<li class="fragment"><p>The notion of principal components as the dimensions that are closest to the <span class="math inline">\(n\)</span> observations extends beyond just the first principal component.</p></li>
<li class="fragment"><p>For instance, the first two principal components of a data set span the plane that is closest to the <span class="math inline">\(n\)</span> observations, in terms of average squared Euclidean distance.</p></li>
</ul>
</section>
<section id="scaling-of-the-variables-matters" class="slide level2 center">
<h2>Scaling of the variables matters</h2>
<ul>
<li class="fragment"><p>If the variables are in different units, scaling each to have standard deviation equal to one is recommended.</p></li>
<li class="fragment"><p>If they are in the same units, you might or might not scale the variables.</p></li>
</ul>

<img data-src="figs/12_4-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="proportion-variance-explained" class="slide level2 center">
<h2>Proportion Variance Explained</h2>
<ul>
<li class="fragment"><p>To understand the strength of each component, we are interested in knowing the proportion of variance explained (PVE) by each one.</p></li>
<li class="fragment"><p>The <em>total variance</em> present in a data set (assuming that the variables have been centered to have mean zero) is defined as</p></li>
</ul>
<p><span class="math display">\[
  \sum_{j=1}^p \text{Var}(X_j) = \sum_{j=1}^p \frac{1}{n} \sum_{i=1}^n x_{ij}^2,
\]</span></p>
<p>and the variance explained by the <span class="math inline">\(m\)</span>-th principal component is</p>
<p><span class="math display">\[
  \text{Var}(Z_m) = \frac{1}{n} \sum_{i=1}^n z_{im}^2.
\]</span></p>
<ul>
<li class="fragment">It can be shown that</li>
</ul>
<p><span class="math display">\[
  \sum_{j=1}^p \text{Var}(X_j) = \sum_{m=1}^M \text{Var}(Z_m),
\]</span></p>
<p>with <span class="math inline">\(M = \min(n-1, p)\)</span>.</p>
</section>
<section id="proportion-variance-explained-continued" class="slide level2 center">
<h2>Proportion Variance Explained: continued</h2>
<ul>
<li class="fragment">Therefore, the PVE of the <span class="math inline">\(m\)</span>-th principal component is given by the positive quantity between 0 and 1:</li>
</ul>
<p><span class="math display">\[
  \frac{\sum_{i=1}^n z_{im}^2}{\sum_{j=1}^p \sum_{i=1}^n x_{ij}^2}.
\]</span></p>
<ul>
<li class="fragment">The PVEs sum to one. We sometimes display the cumulative PVEs.</li>
</ul>

<img data-src="figs/12_3-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="how-many-principal-components-should-we-use" class="slide level2 center">
<h2>How many principal components should we use?</h2>
<p>If we use principal components as a summary of our data, how many components are sufficient?</p>
<ul>
<li class="fragment"><p>No simple answer to this question, as cross-validation is not available for this purpose.</p>
<ul>
<li class="fragment"><p><em>Why not?</em></p></li>
<li class="fragment"><p>When could we use cross-validation to select the number of components?</p></li>
</ul></li>
<li class="fragment"><p>The “scree plot” on the previous slide can be used as a guide: we look for an “elbow”.</p></li>
</ul>
</section>
<section id="clustering" class="slide level2 center">
<h2>Clustering</h2>
<ul>
<li class="fragment"><p><em>Clustering</em> refers to a very broad set of techniques for finding <em>subgroups</em>, or <em>clusters</em>, in a data set.</p></li>
<li class="fragment"><p>We seek a partition of the data into distinct groups so that the observations within each group are quite similar to each other.</p></li>
<li class="fragment"><p>To make this concrete, we must define what it means for two or more observations to be <em>similar</em> or <em>different</em>.</p></li>
<li class="fragment"><p>Indeed, this is often a domain-specific consideration that must be made based on knowledge of the data being studied.</p></li>
</ul>
</section>
<section id="pca-vs-clustering" class="slide level2 center">
<h2>PCA vs Clustering</h2>
<ul>
<li class="fragment"><p>PCA looks for a low-dimensional representation of the observations that explains a good fraction of the variance.</p></li>
<li class="fragment"><p>Clustering looks for homogeneous subgroups among the observations.</p></li>
</ul>
</section>
<section id="clustering-for-market-segmentation" class="slide level2 center">
<h2>Clustering for Market Segmentation</h2>
<ul>
<li class="fragment"><p>Suppose we have access to a large number of measurements (e.g., median household income, occupation, distance from nearest urban area, and so forth) for a large number of people.</p></li>
<li class="fragment"><p>Our goal is to perform <em>market segmentation</em> by identifying subgroups of people who might be more receptive to a particular form of advertising, or more likely to purchase a particular product.</p></li>
<li class="fragment"><p>The task of performing market segmentation amounts to clustering the people in the data set.</p></li>
</ul>
</section>
<section id="two-clustering-methods" class="slide level2 center">
<h2>Two clustering methods</h2>
<ul>
<li class="fragment"><p>In <em>K-means clustering</em>, we seek to partition the observations into a pre-specified number of clusters.</p></li>
<li class="fragment"><p>In <em>hierarchical clustering</em>, we do not know in advance how many clusters we want; in fact, we end up with a tree-like visual representation of the observations, called a <em>dendrogram</em>, that allows us to view at once the clusterings obtained for each possible number of clusters, from 1 to <span class="math inline">\(n\)</span>.</p></li>
</ul>
</section>
<section id="k-means-clustering" class="slide level2 center">
<h2><em>K-means clustering</em></h2>

<img data-src="figs/12_7-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"><p>A simulated data set with 150 observations in 2-dimensional space. Panels show the results of applying <em>K-means clustering</em> with different values of <span class="math inline">\(K\)</span>, the number of clusters. The color of each observation indicates the cluster to which it was assigned using the <em>K-means clustering</em> algorithm. Note that there is no ordering of the clusters, so the cluster coloring is arbitrary. These cluster labels were not used in clustering; instead, they are the outputs of the clustering procedure.</p>
</section>
<section id="details-of-k-means-clustering" class="slide level2 center">
<h2>Details of <em>K-means clustering</em></h2>
<p>Let <span class="math inline">\(C_1, \ldots, C_K\)</span> denote sets containing the indices of the observations in each cluster. These sets satisfy two properties:</p>
<ol type="1">
<li class="fragment"><p><span class="math inline">\(C_1 \cup C_2 \cup \ldots \cup C_K = \{1, \ldots, n\}\)</span>. In other words, each observation belongs to at least one of the <span class="math inline">\(K\)</span> clusters.</p></li>
<li class="fragment"><p><span class="math inline">\(C_k \cap C_{k'} = \emptyset\)</span> for all <span class="math inline">\(k \neq k'\)</span>. In other words, the clusters are non-overlapping: no observation belongs to more than one cluster.</p></li>
</ol>
<p>For instance, if the <span class="math inline">\(i\)</span>-th observation is in the <span class="math inline">\(k\)</span>-th cluster, then <span class="math inline">\(i \in C_k\)</span>.</p>
</section>
<section id="details-of-k-means-clustering-continued" class="slide level2 center">
<h2>Details of <em>K-means clustering</em>: continued</h2>
<ul>
<li class="fragment"><p>The idea behind <em>K-means clustering</em> is that a <strong>good clustering</strong> is one for which the <em>within-cluster variation</em> is as small as possible.</p></li>
<li class="fragment"><p>The within-cluster variation for cluster <span class="math inline">\(C_k\)</span> is a measure <span class="math inline">\(\text{WCV}(C_k)\)</span> of the amount by which the observations within a cluster differ from each other.</p></li>
<li class="fragment"><p>Hence, we want to solve the problem:</p></li>
</ul>
<p><span class="math display">\[
  \text{minimize}_{C_1, \ldots, C_K} \left\{ \sum_{k=1}^K \text{WCV}(C_k) \right\}.
\]</span></p>
<ul>
<li class="fragment">In words, this formula says that we want to partition the observations into <span class="math inline">\(K\)</span> clusters such that the total within-cluster variation, summed over all <span class="math inline">\(K\)</span> clusters, is as small as possible.</li>
</ul>
</section>
<section id="how-to-define-within-cluster-variation" class="slide level2 center">
<h2>How to Define Within-Cluster Variation?</h2>
<ul>
<li class="fragment">Typically, we use Euclidean distance:</li>
</ul>
<p><span class="math display">\[
  \text{WCV}(C_k) = \frac{1}{|C_k|} \sum_{i,i' \in C_k} \sum_{j=1}^p (x_{ij} - x_{i'j})^2, \tag{3}
\]</span></p>
<p>where <span class="math inline">\(|C_k|\)</span> denotes the number of observations in the <span class="math inline">\(k\)</span>-th cluster.</p>
<ul>
<li class="fragment">Combining (2) and (3) gives the optimization problem that defines <em>K-means clustering</em>:</li>
</ul>
<p><span class="math display">\[
  \text{minimize}_{C_1, \ldots, C_K} \left\{ \sum_{k=1}^K \frac{1}{|C_k|} \sum_{i,i' \in C_k} \sum_{j=1}^p (x_{ij} - x_{i'j})^2 \right\}. \tag{4}
\]</span></p>
</section>
<section id="k-means-clustering-algorithm" class="slide level2 center">
<h2>K-Means Clustering Algorithm</h2>
<ol type="1">
<li class="fragment"><p>Randomly assign a number, from 1 to <span class="math inline">\(K\)</span>, to each of the observations. These serve as initial cluster assignments for the observations.</p></li>
<li class="fragment"><p>Iterate until the cluster assignments stop changing:</p>
<p>2.1 For each of the <span class="math inline">\(K\)</span> clusters, compute the cluster <strong>centroid</strong>.<br>
The <span class="math inline">\(k\)</span>-th cluster centroid is the vector of the <span class="math inline">\(p\)</span> feature means for the observations in the <span class="math inline">\(k\)</span>-th cluster.</p>
<p>2.2 Assign each observation to the cluster whose centroid is <strong>closest</strong> (where <strong>closest</strong> is defined using Euclidean distance).</p></li>
</ol>
</section>
<section id="properties-of-the-algorithm" class="slide level2 center">
<h2>Properties of the Algorithm</h2>
<ul>
<li class="fragment">This algorithm is guaranteed to decrease the value of the objective (4) at each step. <strong>Why?</strong> Note that</li>
</ul>
<p><span class="math display">\[
  \frac{1}{|C_k|} \sum_{i,i' \in C_k} \sum_{j=1}^p (x_{ij} - x_{i'j})^2 = 2 \sum_{i \in C_k} \sum_{j=1}^p (x_{ij} - \bar{x}_{kj})^2,
\]</span></p>
<p>where <span class="math inline">\(\bar{x}_{kj} = \frac{1}{|C_k|} \sum_{i \in C_k} x_{ij}\)</span> is the mean for feature <span class="math inline">\(j\)</span> in cluster <span class="math inline">\(C_k\)</span>.</p>
<ul>
<li class="fragment">However, it is not guaranteed to give the global minimum. <strong>Why not?</strong></li>
</ul>
</section>
<section id="example" class="slide level2 center">
<h2>Example</h2>

<img data-src="figs/12_8-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="details-of-previous-figure" class="slide level2 center">
<h2>Details of Previous Figure</h2>
<ul>
<li class="fragment"><p><strong>Top left</strong>: The observations are shown.</p></li>
<li class="fragment"><p><strong>Top center</strong>: In Step 1 of the algorithm, each observation is randomly assigned to a cluster.</p></li>
<li class="fragment"><p><strong>Top right</strong>: In Step 2(a), the cluster centroids are computed. These are shown as large colored disks. Initially, the centroids are almost completely overlapping because the initial cluster assignments were chosen at random.</p></li>
<li class="fragment"><p><strong>Bottom left</strong>: In Step 2(b), each observation is assigned to the nearest centroid.</p></li>
<li class="fragment"><p><strong>Bottom center</strong>: Step 2(a) is once again performed, leading to new cluster centroids.</p></li>
<li class="fragment"><p><strong>Bottom right</strong>: The results obtained after 10 iterations.</p></li>
</ul>
</section>
<section id="example-different-starting-values" class="slide level2 center">
<h2>Example: Different Starting Values</h2>

<img data-src="figs/12_9-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="details-of-previous-figure-1" class="slide level2 center">
<h2>Details of Previous Figure</h2>
<ul>
<li class="fragment"><p><strong>K-means clustering</strong> was performed six times on the data from the previous figure with <span class="math inline">\(K = 3\)</span>, each time with a different random assignment of the observations in Step 1 of the K-means algorithm.</p></li>
<li class="fragment"><p>Above each plot is the value of the objective (4).</p></li>
<li class="fragment"><p>Three different local optima were obtained, one of which resulted in a smaller value of the objective and provides better separation between the clusters.</p></li>
<li class="fragment"><p>Those labeled in <strong>red</strong> all achieved the same best solution, with an objective value of <strong>235.8</strong>.</p></li>
</ul>
</section>
<section id="hierarchical-clustering" class="slide level2 center">
<h2>Hierarchical Clustering</h2>
<ul>
<li class="fragment"><p><strong>K-means clustering</strong> requires us to pre-specify the number of clusters <span class="math inline">\(K\)</span>. This can be a disadvantage (later we discuss strategies for choosing <span class="math inline">\(K\)</span>).</p></li>
<li class="fragment"><p><strong>Hierarchical clustering</strong> is an alternative approach which does not require that we commit to a particular choice of <span class="math inline">\(K\)</span>.</p></li>
<li class="fragment"><p>In this section, we describe <strong>bottom-up</strong> or <strong>agglomerative clustering</strong>. This is the most common type of hierarchical clustering and refers to the fact that a dendrogram is built starting from the leaves and combining clusters up to the trunk.</p></li>
</ul>
</section>
<section id="hierarchical-clustering-the-idea" class="slide level2 center">
<h2>Hierarchical Clustering: The Idea</h2>
<p>Builds a hierarchy in a “bottom-up” fashion…</p>
<p>Step 1</p>

<img data-src="figs/12_1_3-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="hierarchical-clustering-the-idea-step-2" class="slide level2 center">
<h2>Hierarchical Clustering: The Idea (Step 2)</h2>
<p>Merging Closest Observations</p>

<img data-src="figs/12_1_4-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="hierarchical-clustering-the-idea-step-3" class="slide level2 center">
<h2>Hierarchical Clustering: The Idea (Step 3)</h2>
<p>Merging Another Closest Pair</p>

<img data-src="figs/12_1_5-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="hierarchical-clustering-the-idea-step-4" class="slide level2 center">
<h2>Hierarchical Clustering: The Idea (Step 4)</h2>
<p>Expanding the Hierarchy</p>

<img data-src="figs/12_1_6-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="hierarchical-clustering-the-idea-final-step" class="slide level2 center">
<h2>Hierarchical Clustering: The Idea (Final Step)</h2>
<p>Single Cluster Representation</p>

<img data-src="figs/12_1_7-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="hierarchical-clustering-algorithm" class="slide level2 center">
<h2>Hierarchical Clustering Algorithm</h2>
<p>The approach in words:</p>
<ul>
<li class="fragment"><p>Start with each point in its own cluster.</p></li>
<li class="fragment"><p>Identify the closest two clusters and merge them.</p></li>
<li class="fragment"><p>Repeat.</p></li>
<li class="fragment"><p>Ends when all points are in a single cluster.</p></li>
</ul>

<img data-src="figs/12_1_8-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="an-example" class="slide level2 center">
<h2>An Example</h2>

<img data-src="figs/12_10-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"><ul>
<li class="fragment"><p>45 observations generated in 2-dimensional space.</p></li>
<li class="fragment"><p>In reality, there are three distinct classes, shown in separate colors.</p></li>
<li class="fragment"><p>However, we will treat these class labels as unknown and will seek to cluster the observations in order to discover the classes from the data.</p></li>
</ul>
</section>
<section id="application-of-hierarchical-clustering" class="slide level2 center">
<h2>Application of Hierarchical Clustering</h2>

<img data-src="figs/12_11-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"><ul>
<li class="fragment"><p>This slide visualizes the application of hierarchical clustering on data.</p></li>
<li class="fragment"><p>Each dendrogram represents different stages or thresholds for determining clusters.</p></li>
<li class="fragment"><p>The dashed horizontal lines indicate where the data is split into distinct clusters.</p></li>
</ul>
</section>
<section id="details-of-previous-figure-2" class="slide level2 center">
<h2>Details of Previous Figure</h2>
<ul>
<li class="fragment"><p><strong>Left</strong>: Dendrogram obtained from hierarchically clustering the data from the previous slide, using complete linkage and Euclidean distance.</p></li>
<li class="fragment"><p><strong>Center</strong>: The dendrogram from the left-hand panel, cut at a height of 9 (indicated by the dashed line). This cut results in two distinct clusters, shown in different colors.</p></li>
<li class="fragment"><p><strong>Right</strong>: The dendrogram from the left-hand panel, now cut at a height of 5. This cut results in three distinct clusters, shown in different colors. Note that the colors were not used in clustering, but are simply used for display purposes in this figure.</p></li>
</ul>
</section>
<section id="types-of-linkage" class="slide level2 center">
<h2>Types of Linkage</h2>
<table class="caption-top">
<colgroup>
<col style="width: 43%">
<col style="width: 56%">
</colgroup>
<thead>
<tr class="header">
<th><strong>Linkage</strong></th>
<th><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Complete</strong></td>
<td>Maximal inter-cluster dissimilarity. Compute all pairwise dissimilarities between the observations in cluster A and the observations in cluster B, and record the <em>largest</em> of these dissimilarities.</td>
</tr>
<tr class="even">
<td><strong>Single</strong></td>
<td>Minimal inter-cluster dissimilarity. Compute all pairwise dissimilarities between the observations in cluster A and the observations in cluster B, and record the <em>smallest</em> of these dissimilarities.</td>
</tr>
<tr class="odd">
<td><strong>Average</strong></td>
<td>Mean inter-cluster dissimilarity. Compute all pairwise dissimilarities between the observations in cluster A and the observations in cluster B, and record the <em>average</em> of these dissimilarities.</td>
</tr>
<tr class="even">
<td><strong>Centroid</strong></td>
<td>Dissimilarity between the centroid for cluster A (a mean vector of length <em>p</em>) and the centroid for cluster B. Centroid linkage can result in undesirable <em>inversions</em>.</td>
</tr>
</tbody>
</table>
</section>
<section id="choice-of-dissimilarity-measure" class="slide level2 center">
<h2>Choice of Dissimilarity Measure</h2>
<ul>
<li class="fragment"><p>So far, have used Euclidean distance.</p></li>
<li class="fragment"><p>An alternative is <em>correlation-based distance</em>, which considers two observations to be similar if their features are highly correlated.</p></li>
<li class="fragment"><p>This is an unusual use of correlation, which is normally computed between variables; here, it is computed between the observation profiles for each pair of observations.</p></li>
</ul>

<img data-src="figs/12_15-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="practical-issues" class="slide level2 center">
<h2>Practical Issues</h2>
<ul>
<li class="fragment"><p><em>Scaling of the variables matters!</em> Should the observations or features first be standardized in some way?</p>
<ul>
<li class="fragment">For instance, maybe the variables should be centered to have mean zero and scaled to have a standard deviation of one.</li>
</ul></li>
<li class="fragment"><p>In the case of hierarchical clustering:</p>
<ul>
<li class="fragment"><p>What dissimilarity measure should be used?</p></li>
<li class="fragment"><p>What type of linkage should be used?</p></li>
</ul></li>
<li class="fragment"><p>How many clusters to choose? (in both <em>K</em>-means or hierarchical clustering):</p>
<ul>
<li class="fragment"><p>Difficult problem.</p></li>
<li class="fragment"><p>No agreed-upon method. See <em>Elements of Statistical Learning</em>, Chapter 13, for more details.</p></li>
</ul></li>
<li class="fragment"><p>Which features should we use to drive the clustering?</p></li>
</ul>
</section>
<section id="example-breast-cancer-microarray-study" class="slide level2 center">
<h2>Example: Breast Cancer Microarray Study</h2>
<ul>
<li class="fragment"><p><strong>“Repeated observation of breast tumor subtypes in independent gene expression data sets;”</strong> Sorlie et al., PNAS 2003.</p></li>
<li class="fragment"><p>Gene expression measurements for approximately ~8000 genes, for each of 88 breast cancer patients.</p></li>
<li class="fragment"><p>Average linkage, correlation metric.</p></li>
<li class="fragment"><p>Clustered samples using <strong>500 intrinsic genes</strong>:</p>
<ul>
<li class="fragment"><p>Each woman was measured before and after chemotherapy.</p></li>
<li class="fragment"><p>Intrinsic genes have the smallest within/between variation.</p></li>
</ul></li>
</ul>
</section>
<section id="section" class="slide level2 center">
<h2></h2>

<img data-src="figs/12_1_9-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="section-1" class="slide level2 center">
<h2></h2>

<img data-src="figs/12_1_10-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="conclusions" class="slide level2 center">
<h2>Conclusions</h2>
<ul>
<li class="fragment"><p><em>Unsupervised learning</em> is important for understanding the variation and grouping structure of a set of unlabeled data, and can be a useful pre-processor for supervised learning.</p></li>
<li class="fragment"><p>It is intrinsically more difficult than <em>supervised learning</em> because there is no gold standard (like an outcome variable) and no single objective (like test set accuracy).</p></li>
<li class="fragment"><p>It is an active field of research, with many recently developed tools such as <em>self-organizing maps</em>, <em>independent components analysis</em>, and <em>spectral clustering</em>.</p></li>
<li class="fragment"><p>See <em>The Elements of Statistical Learning</em>, chapter 14.</p></li>
</ul>
</section></section>
<section>
<section id="matrix-completion-and-missing-values" class="title-slide slide level1 center">
<h1>Matrix Completion and Missing Values</h1>

</section>
<section id="matrix-completion-and-missing-values-1" class="slide level2 center">
<h2>Matrix Completion and Missing Values</h2>
<ul>
<li class="fragment"><p>It is often the case that data matrices <strong>X</strong> have missing entries, often represented by <strong>NAs</strong> (not available).</p></li>
<li class="fragment"><p>This is a nuisance, since many of our modeling procedures, such as linear regression and GLMs, require complete data.</p></li>
<li class="fragment"><p>Sometimes imputation <em>is the prediction problem!</em> — as in recommender systems.</p></li>
<li class="fragment"><p>One simple approach is <em>mean imputation</em> — replace missing values for a variable by the mean of the non-missing entries.</p></li>
<li class="fragment"><p>This ignores the correlations among variables; we should be able to exploit these correlations when imputing missing values.</p></li>
<li class="fragment"><p>We assume values are missing <em>at random</em>; i.e., the missingness should not be informative.</p></li>
<li class="fragment"><p>We present an approach based on principal components.</p></li>
</ul>
</section>
<section id="recommender-systems" class="slide level2 center">
<h2>Recommender Systems</h2>

<img data-src="figs/12_1_11-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"></section>
<section id="recommender-systems-1" class="slide level2 center">
<h2>Recommender Systems</h2>
<table class="caption-top">
<colgroup>
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 4%">
<col style="width: 11%">
<col style="width: 10%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 8%">
<col style="width: 8%">
<col style="width: 4%">
<col style="width: 11%">
<col style="width: 2%">
</colgroup>
<thead>
<tr class="header">
<th>Customer</th>
<th>Jerry Maguire</th>
<th>Oceans</th>
<th>Road to Perdition</th>
<th>A Fortunate Man</th>
<th>Catch Me If You Can</th>
<th>Driving Miss Daisy</th>
<th>The Two Popes</th>
<th>The Laundromat</th>
<th>Code 8</th>
<th>The Social Network</th>
<th>…</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Customer 1</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>4</td>
<td>.</td>
<td>.</td>
<td>…</td>
</tr>
<tr class="even">
<td>Customer 2</td>
<td>.</td>
<td>3</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>3</td>
<td>.</td>
<td>3</td>
<td>…</td>
</tr>
<tr class="odd">
<td>Customer 3</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>2</td>
<td>4</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>…</td>
</tr>
<tr class="even">
<td>Customer 4</td>
<td>3</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>2</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>…</td>
</tr>
<tr class="odd">
<td>Customer 5</td>
<td>5</td>
<td>1</td>
<td>.</td>
<td>.</td>
<td>4</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>…</td>
</tr>
<tr class="even">
<td>Customer 6</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>2</td>
<td>.</td>
<td>.</td>
<td>4</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>…</td>
</tr>
<tr class="odd">
<td>Customer 7</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>5</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>3</td>
<td>.</td>
<td>…</td>
</tr>
<tr class="even">
<td>Customer 8</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>…</td>
</tr>
<tr class="odd">
<td>Customer 9</td>
<td>3</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>5</td>
<td>.</td>
<td>1</td>
<td>…</td>
</tr>
<tr class="even">
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
</tbody>
</table>
<ul>
<li class="fragment"><p>Netflix users rate movies they have seen, usually a very small fraction of available movies.</p></li>
<li class="fragment"><p>Predicting missing ratings provides a way to <em>recommend</em> movies to users. <strong>Matrix completion</strong> is one of the primary tools.</p></li>
</ul>
</section>
<section id="matrix-approximation-via-principal-components" class="slide level2 center">
<h2>Matrix Approximation via Principal Components</h2>
<ul>
<li class="fragment">In Section 12.2.2 we gave an interpretation of principal components in terms of <em>matrix approximation</em>:</li>
</ul>
<p><span class="math display">\[
\text{minimize}_{\mathbf{A} \in \mathbb{R}^{n \times M}, \mathbf{B} \in \mathbb{R}^{p \times M}} \left\{ \sum_{j=1}^p \sum_{i=1}^n \left( x_{ij} - \sum_{m=1}^M a_{im}b_{jm} \right)^2 \right\}.
\]</span></p>
<ul>
<li class="fragment"><p><strong><span class="math inline">\(\mathbf{A}\)</span></strong> is an <span class="math inline">\(n \times M\)</span> matrix whose <span class="math inline">\((i, m)\)</span> element is <span class="math inline">\(a_{im}\)</span>, and <strong><span class="math inline">\(\mathbf{B}\)</span></strong> is a <span class="math inline">\(p \times M\)</span> matrix whose <span class="math inline">\((j, m)\)</span> element is <span class="math inline">\(b_{jm}\)</span>.</p></li>
<li class="fragment"><p>It can be shown that for any value of <span class="math inline">\(M\)</span>, the <strong>first <span class="math inline">\(M\)</span> principal components</strong> provide a solution: <span class="math display">\[
\hat{a}_{im} = z_{im} \quad \text{and} \quad \hat{b}_{jm} = \phi_{jm}.
\]</span></p></li>
<li class="fragment"><p>But what to do if the matrix has missing elements?</p></li>
</ul>
</section>
<section id="matrix-completion-via-principal-components" class="slide level2 center">
<h2>Matrix Completion via Principal Components</h2>
<p>We pose instead a modified version of the approximation criterion:</p>
<p><span class="math display">\[
\text{minimize}_{\mathbf{A} \in \mathbb{R}^{n \times M}, \mathbf{B} \in \mathbb{R}^{p \times M}} \left\{ \sum_{(i,j) \in \mathcal{O}} \left( x_{ij} - \sum_{m=1}^M a_{im}b_{jm} \right)^2 \right\},
\]</span></p>
<p>where <span class="math inline">\(\mathcal{O}\)</span> is the set of all <strong>observed</strong> pairs of indices <span class="math inline">\((i, j)\)</span>, a subset of the possible <span class="math inline">\(n \times p\)</span> pairs.</p>
<p>Once we solve this problem:</p>
<ul>
<li class="fragment">We can estimate a missing observation <span class="math inline">\(x_{ij}\)</span> using:</li>
</ul>
<p><span class="math display">\[
  \hat{x}_{ij} = \sum_{m=1}^M \hat{a}_{im}\hat{b}_{jm},
\]</span></p>
<p>where <span class="math inline">\(\hat{a}_{im}\)</span> and <span class="math inline">\(\hat{b}_{jm}\)</span> are the <span class="math inline">\((i,m)\)</span> and <span class="math inline">\((j,m)\)</span> elements of the solution matrices <span class="math inline">\(\mathbf{\hat{A}}\)</span> and <span class="math inline">\(\mathbf{\hat{B}}\)</span>.</p>
<ul>
<li class="fragment">We can (approximately) recover the <span class="math inline">\(M\)</span> principal component scores and loadings, as if data were complete.</li>
</ul>
</section>
<section id="iterative-algorithm-for-matrix-completion" class="slide level2 center">
<h2>Iterative Algorithm for Matrix Completion</h2>
<ol type="1">
<li class="fragment"><p><strong>Initialize</strong>: create a complete data matrix <span class="math inline">\(\tilde{\mathbf{X}}\)</span> by filling in the missing values using mean imputation.</p></li>
<li class="fragment"><p><strong>Repeat</strong>: steps (a)–(c) until the objective in (c) fails to decrease:</p>
<ul>
<li class="fragment"><ol type="a">
<li class="fragment"></li>
</ol></li>
</ul></li>
</ol>
<p><span class="math display">\[
      \text{minimize}_{\mathbf{A} \in \mathbb{R}^{n \times M}, \mathbf{B} \in \mathbb{R}^{p \times M}} \left\{ \sum_{j=1}^p \sum_{i=1}^n \left( \tilde{x}_{ij} - \sum_{m=1}^M a_{im}b_{jm} \right)^2 \right\},
\]</span></p>
<pre><code>  by computing the principal components of $\tilde{\mathbf{X}}$.

- (b) For each missing entry $(i, j) \notin \mathcal{O}$, set:</code></pre>
<p><span class="math display">\[
      \tilde{x}_{ij} \leftarrow \sum_{m=1}^M \hat{a}_{im} \hat{b}_{jm}.
\]</span></p>
<pre><code>- (c) Compute the objective:</code></pre>
<p><span class="math display">\[
      \sum_{(i,j) \in \mathcal{O}} \left( x_{ij} - \sum_{m=1}^M \hat{a}_{im} \hat{b}_{jm} \right)^2.
\]</span></p>
<ol start="3" type="1">
<li class="fragment"><strong>Return</strong> the estimated missing entries <span class="math inline">\(\tilde{x}_{ij}\)</span>, <span class="math inline">\((i, j) \notin \mathcal{O}\)</span>.</li>
</ol>
</section>
<section id="example-usaarrests-data" class="slide level2 center">
<h2>Example: USAarrests Data</h2>

<img data-src="figs/12_1_12-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:65.0%"><p>Here <span class="math inline">\(\mathbf{X}\)</span> has 50 rows (states) and four columns: <strong>Murder</strong>, <strong>Assault</strong>, <strong>Rape</strong>, and <strong>UrbanPop</strong>.</p>
<ul>
<li class="fragment"><p>We selected 20 states at random, and for each, we selected one of the variables at random, and set its value to <strong>NA</strong>.</p></li>
<li class="fragment"><p>Used <span class="math inline">\(M = 1\)</span> principal component in the algorithm.</p></li>
<li class="fragment"><p><strong>Correlation</strong>: 0.63 between original and imputed values.</p></li>
</ul>
</section>
<section id="example-continued" class="slide level2 center">
<h2>Example — Continued</h2>
<ul>
<li class="fragment"><p>The <strong>USAarrests</strong> data has only four variables, which is on the low end for this method to work well. For this reason, for this demonstration we randomly set at most one variable per state to be missing and only used <span class="math inline">\(M = 1\)</span> principal component.</p></li>
<li class="fragment"><p>In general, in order to apply this algorithm, we must select <span class="math inline">\(M\)</span>, the number of principal components to use for the imputation.</p></li>
<li class="fragment"><p>One approach is to randomly set to <strong>NA</strong> some elements that were actually observed, and select <span class="math inline">\(M\)</span> based on how well those known values are recovered. This is closely related to the validation-set approach seen in Chapter 5.</p></li>
<li class="fragment"><p><strong>softImpute</strong> package in R implements matrix completion algorithms and can manage Netflix-scale matrices.</p></li>
</ul>
</section>
<section id="summary" class="slide level2 center">
<h2>Summary</h2>
<div>
<div class="columns">
<div class="column" style="width:50%;">
<div style="font-size: 80%;">
<ul>
<li>XXXX</li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<div style="font-size: 80%;">
<ul>
<li>XXXX</li>
</ul>
</div>
</div></div>
</div>
</section></section>
<section id="thank-you" class="title-slide slide level1 center" data-background-color="#cfb991">
<h1>Thank you!</h1>

<div class="quarto-auto-generated-content">
<div class="footer footer-default">
<p>Predictive Analytics</p>
</div>
</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': true,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":false},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: true,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'slide',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'fade',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1600,

        height: 900,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        // For code content inside modals, clipBoardJS needs to be initialized with a container option
        // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>